### PR TITLE
VM: Split up IViewManager into optional interfaces #3565

### PR DESCRIPTION
**Pull request by jonthysell**
This makes it easier to only implement the ViewManager features you need.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3598)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/IViewManager.idl [hunk 1]
```diff
@@ -19,13 +19,25 @@ namespace Microsoft.ReactNative.Bridge
...
+  {
     Windows.Foundation.Collections.IMapView<String, IInspectable> ExportedViewConstants { get; };
+  }
 
+  [webhosthidden]
+  interface IViewManagerWithNativeProperties
+  {
     Windows.Foundation.Collections.IMapView<String, ViewManagerPropertyType> NativeProps { get; };
 
     void UpdateProperties(Windows.UI.Xaml.FrameworkElement view, Windows.Foundation.Collections.IMapView<String, IInspectable> propertyMap);
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/IViewManager.idl]: Should it be NativeProps and UpdateProps? Or NativeProperties and UpdateProperties? Them being mismatched seems strange.

**Comment by vmoroz** [comment]
Could we change it instead towards the builder pattern as we do for the native modules?

File: vnext/Microsoft.ReactNative.SharedManaged/AttributedViewManager.cs [hunk 2]
```diff
@@ -4,15 +4,19 @@
...
 using Microsoft.ReactNative.Bridge;
 
 namespace Microsoft.ReactNative.Managed
 {
-  internal abstract class AttributedViewManager<T> : IViewManager where T : FrameworkElement, new()
+  internal abstract class AttributedViewManager<T> :
+    IViewManager,
+    IViewManagerWithExportedViewConstants,
+    IViewManagerWithNativeProperties,
+    IViewManagerWithCommands
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative.SharedManaged/AttributedViewManager.cs]: I have mixed feelings about this pattern. It makes implementing both more simple and more fussy at the same time. I think on the whole it's a net positive but it's probably worth floating some discussion on API review. There may be some other patterns we can adopt.

**Comment by jonthysell** [comment]
> Could we change it instead towards the builder pattern as we do for the native modules?

That was one option, but looking into it, I don't think it's worth the effort. ViewManagers have a lot of specific contracts, and the benefits are pretty minimal to building a whole builder infrastructure for them. Breaking up the interfaces gives plenty of opt-in feature support.

**Pull request merged**